File: src/dev/wirlie/bungeecord/glist/activity/ActivityManager.java
package dev.wirlie.bungeecord.glist.activity;

import dev.wirlie.bungeecord.glist.TemporalPaginator;
import dev.wirlie.bungeecord.glist.executor.GlistCommand;
import dev.wirlie.bungeecord.glist.util.PlayerGlistEntry;
import org.bukkit.entity.Player;

import java.util.*;

public class ActivityManager {

    private final Map<UUID, Set<ActivityType>> activitiesByPlayer = new HashMap<>();

    public void addActivity(UUID player, ActivityType activityType) {
        Set<ActivityType> activities = activitiesByPlayer.computeIfAbsent(player, k -> new HashSet<>());
        activities.add(activityType);

        synchronized (GlistCommand.serversPaginators) {
            for (TemporalPaginator paginator : GlistCommand.serversPaginators.values()) {
                for (PlayerGlistEntry entry : paginator.getFullData()) {
                    if (entry.getPlayer().getUniqueId().equals(player)) {
                        entry.getActivities().add(activityType);
                    }
                }
            }
        }
    }

    public Set<ActivityType> getActivities(UUID player) {
        Set<ActivityType> activities = activitiesByPlayer.get(player);
        if(activities == null) {
            return new HashSet<>();
        }
        return activities;
    }

    public void removeActivity(UUID player, ActivityType activityType) {
        Set<ActivityType> activities = activitiesByPlayer.get(player);
        if(activities == null) {
            return;
        }

        activities.remove(activityType);

        if(activities.size() == 0) {
            activitiesByPlayer.remove(player);
        }

        synchronized (GlistCommand.serversPaginators) {
            for (TemporalPaginator paginator : GlistCommand.serversPaginators.values()) {
                for (PlayerGlistEntry entry : paginator.getFullData()) {
                    if (entry.getPlayer().getUniqueId().equals(player)) {
                        entry.getActivities().remove(activityType);
                    }
                }
            }
        }
    }

    public void removePlayer(UUID player) {
        activitiesByPlayer.remove(player);
        synchronized (GlistCommand.serversPaginators) {
            for (TemporalPaginator paginator : GlistCommand.serversPaginators.values()) {
                for (PlayerGlistEntry entry : paginator.getFullData()) {
                    if (entry.getPlayer().getUniqueId().equals(player)) {
                        entry.getActivities().clear();
                    }
                }
            }
        }
    }

}

File: Platform-BungeeCord/src/main/kotlin/dev/wirlie/bungeecord/glist/activity/ActivityManager.kt
package dev.wirlie.bungeecord.glist.activity

import dev.wirlie.bungeecord.glist.executor.GlistCommand
import java.util.*

class ActivityManager {

    private val activitiesByPlayer = Collections.synchronizedMap(HashMap<UUID, MutableSet<ActivityType>>())

    fun addActivity(player: UUID, activityType: ActivityType) {
        val activities = activitiesByPlayer.computeIfAbsent(player) { HashSet() }
        activities.add(activityType)
        synchronized(GlistCommand.serversPaginators) {
            for (paginator in GlistCommand.serversPaginators.values) {
                for (entry in paginator.fullData) {
                    if (entry.player.uniqueId == player) {
                        entry.activities.add(activityType)
                    }
                }
            }
        }
    }

    fun getActivities(player: UUID): MutableSet<ActivityType> {
        return activitiesByPlayer[player]
            ?: return HashSet()
    }

    fun removeActivity(player: UUID, activityType: ActivityType) {
        val activities = activitiesByPlayer[player] ?: return
        activities.remove(activityType)
        if (activities.size == 0) {
            activitiesByPlayer.remove(player)
        }
        synchronized(GlistCommand.serversPaginators) {
            for (paginator in GlistCommand.serversPaginators.values) {
                for (entry in paginator.fullData) {
                    if (entry.player.uniqueId == player) {
                        entry.activities.remove(activityType)
                    }
                }
            }
        }
    }

    fun removePlayer(player: UUID) {
        activitiesByPlayer.remove(player)
        synchronized(GlistCommand.serversPaginators) {
            for (paginator in GlistCommand.serversPaginators.values) {
                for (entry in paginator.fullData) {
                    if (entry.player.uniqueId == player) {
                        entry.activities.clear()
                    }
                }
            }
        }
    }
}
